### ISSUE TITLE
[Arcade v6] Remove stable version restriction from non-isolated feeds

### ISSUE THREAD

**Issue by brettfo** [dotnet/arcade#6616]
Similar to issue #5487, the stable package version restriction in non-isolated feeds needs to be removed for v6.  The dotnet/command-line-api repo is on Arcade v6 but fails during package publish for the reasons described in the previously mentioned issue.

Publish failure is [here](https://dev.azure.com/dnceng/internal/_build/results?buildId=898680&view=logs&j=0b850c37-f3ae-5bb9-a9fb-4f1e0fce2747&t=50d41e72-ce5e-50a3-efc7-c28c030be91c).

This issue is blocking since we can no longer get around this by publishing to MyGet.

**Comment by brettfo** [comment]
Not sure if it's relevant, but after updating the command-line-api repo to v3 package publishing, [the publish step is now failing](https://dev.azure.com/dnceng/internal/_build/results?buildId=898875&view=logs&j=226748d0-f812-5437-d3f0-2dd291f5666e&t=bad11196-972e-5d03-45a8-9db526506073) with:

```
The promotion build finished but the build isn't associated with at least one of the target channels. This is an error scenario. Please contact @dnceng.
```

**Comment by markwilkie** [comment]
cc/ @epananth @riarenas

**Comment by riarenas** [comment]
The error in the promotion build from the update to v3 publishing is the same: ```Package 'dotnet-suggest' has stable version '1.1.157406' but is targeted at a non-isolated feed```

@epananth This looks similar to what you helped @JoeRobich recently. The validations now have an opt-out mechanism that should automatically happen if the `AutoGenerateAssemblyVersion` msbuild property is set to true and  `PreReleaseVersionLabel` property is empty.

**Comment by epananth** [comment]
@riarenas you are right 

We need to have

1) PreReleaseVersionLabel empty and 
2) AutoGenerateAssemblyVersion = true 

Right now PreReleaseVersionLabel has value beta -> https://dev.azure.com/dnceng/internal/_git/dotnet-command-line-api?path=%2Feng%2FVersions.props&version=GBmain&line=11&lineEnd=12&lineStartColumn=1&lineEndColumn=1&lineStyle=plain&_a=contents

AutoGenerateAssemblyVersion is set to false => https://dev.azure.com/dnceng/internal/_git/dotnet-command-line-api?path=%2Feng%2FVersions.props&version=GBmain&line=23&lineEnd=24&lineStartColumn=1&lineEndColumn=1&lineStyle=plain&_a=contents

Once this is updated publishing should work. 

@brettfo let me know if you have any questions..

**Comment by epananth** [comment]
@brettfo I checked the post build logs and looks like the UseGlobalToolVersion is not set to true and the PreReleaseVersionLable is still using beta

Snapshot from the build 
![image](https://user-images.githubusercontent.com/47157394/100697906-105d1980-334c-11eb-85aa-8d85d93b4348.png)

**Comment by epananth** [comment]
This is where you need to change it to use 'UseGlobalToolVersion'

https://dev.azure.com/dnceng/internal/_git/dotnet-command-line-api?path=%2Fsrc%2FSystem.CommandLine.Tests%2FDirectory.Build.props&version=GBmain&line=4&lineEnd=5&lineStartColumn=1&lineEndColumn=1&lineStyle=plain&_a=contents

**Comment by epananth** [comment]
Here https://github.com/dotnet/command-line-api/blob/main/src/System.CommandLine/Directory.Build.props

**Comment by brettfo** [comment]
I've experimented some more and it seems that the regression is more wide-spread than initially thought.  I removed the v3 package publishing from the dotnet/interactive build, but it still fails on package publish for the [same reasons](https://dev.azure.com/dnceng/internal/_build/results?buildId=904632&view=logs&j=51c75269-d108-5b35-500b-cc803a6d6b66&t=db467048-79f0-5bc7-fe0b-67cface4c44b&l=57).

In looking through the old issues and the Arcade code, PR #4682 removed the stable version check, but the current state of the [V2](https://github.com/dotnet/arcade/blob/6d8efa00a7dd2d15e07df673a83cecb8a0a3031f/src/Microsoft.DotNet.Build.Tasks.Feed/src/PublishArtifactsInManifestV2.cs#L61) and [V3](https://github.com/dotnet/arcade/blob/6d8efa00a7dd2d15e07df673a83cecb8a0a3031f/src/Microsoft.DotNet.Build.Tasks.Feed/src/PublishArtifactsInManifestV3.cs#L149) code still performs this check.

Given that we have no rollback, this issue is currently blocking all package publish from dotnet/command-line-api, dotnet/interactive, and very soon dotnet/fsharp.

**Comment by epananth** [comment]
Discussed with @mmitche and @riarenas 

Previously we introduced a change to skip the safety checks by introducing IsReleaseOnlyPackage, this was dependent on PreReleaseVersionLabel version and AutoGenerateAssemblyVersion. 

@bret here is publishing to partly stable and partly beta versions. 

Fix for this -> Making IsReleaseOnlyPackage settable by the repo owners in version.props so it will work in this scenario. Moving this to FR.

**Comment by mmitche** [comment]
@brettfo The stable version check is pretty critical and I think we need to go with @epananth's fix. It is not really a regression. It was removed a while back because of a repo that was producing non-prerelease versions that did in fact always vary build to build. I decided at the time it was unlikely to cause a problem if the check was removed.

This turned out to be a bad idea. When we were doing GA testing, some repos were erroneously producing stable versions for non-shipping packages, which then got shuttled off to non-isolated feeds and potentially could have caused major issues.

Having an override at your repo level to say: "i know what I'm doing" is the right way to go here, but disabling the check altogether is too dangerous.

**Comment by brettfo** [comment]
Is the override to set `$(IsReleaseOnlyPackage)` to `true` in our `eng/Versions.props` file?  If that's set will that still enable us to produce and publish `*-beta.12345` packages, too?  And finally, if that is the fix, is that something we can do now, or does it require an update to Arcade then taking those changes into our repo?

**Comment by mmitche** [comment]
I don't think the override currently exists. Basically the override should probably be "should the checks be ignored". Right now the checks are ignored based on inferring the value.

@brettfo You may be able to work around this by inverting your properties for now. In eng/Versions.props set:

PreReleaseVersionLabel empty and
AutoGenerateAssemblyVersion = true

and then in each project that doesn't use that, set PreReleaseVersionLabel to beta and AutoGenerateAssemblyVersion to false.

**Comment by brettfo** [comment]
@mmitche @epananth Can either of you check out PR dotnet/command-line-api#1113?  I did the property inversion as suggested, but to keep the churn small and so it can be easily reverted, I've kept all changes to `eng/Versions.props`.  In that PR I set `AutoGenerateAssemblyVersion` to `true` then never set it to false; it didn't seem to affect anything negatively, am I reading that right?

**Comment by brettfo** [comment]
Following the suggestions provided, I was able to unblock those builds.

Now I'm seeing a related, but slightly different issue in dotnet/fsharp.  The signed builds from `main` are passing just fine, but our [builds from `release/dev16.9`](https://dev.azure.com/dnceng/internal/_build/results?buildId=908425&view=results) are [failing on package publish](https://dnceng.visualstudio.com/internal/_build/results?buildId=908496&view=logs&j=ba23343f-f710-5af9-782d-5bd26b102304&t=6e277ba4-1c1e-552d-b96f-db0aeb4be20a&l=31) with:

```
MSBUILD : error MSB1006: Property is not valid.
Switch: 1473
```

**Comment by epananth** [comment]
This build was trying to add the build to 2 channels and was failing, so in order to unblock you , I added the build to channel individually 

https://dnceng.visualstudio.com/internal/_build/results?buildId=913400&view=results

https://dnceng.visualstudio.com/internal/_build/results?buildId=913324&view=logs&s=380ab8c6-4139-55ff-15c4-fb48260f4dca

This is just to unblock you for now. We can delete one of the default channels for now then when the build runs 
(say you delete VS master)

1) Build will use the default channel VS 16.9 
2) We can run darc add-build-to-channel for the VS master channel 

** Note this is just a work around. I will try to dig more on this tomorrow.

**Comment by epananth** [comment]
Created another issue to track the f sharp build error https://github.com/dotnet/arcade/issues/6669

**Comment by MattGal** [comment]
Changes rolled out, closing.
